File: system/platform-core/src/main/java/org/platformlambda/core/actuator/HealthService.java
/*

    Copyright 2018-2022 Accenture Technology

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.

 */

package org.platformlambda.core.actuator;

import org.platformlambda.core.exception.AppException;
import org.platformlambda.core.models.EventEnvelope;
import org.platformlambda.core.models.Kv;
import org.platformlambda.core.models.LambdaFunction;
import org.platformlambda.core.system.Platform;
import org.platformlambda.core.system.PostOffice;
import org.platformlambda.core.util.AppConfigReader;
import org.platformlambda.core.util.ManagedCache;
import org.platformlambda.core.util.Utility;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeoutException;

public class HealthService implements LambdaFunction {
    private static final Logger log = LoggerFactory.getLogger(HealthService.class);

    private static final ManagedCache cache = ManagedCache.createCache("health.info", 5000);
    private static final String TYPE = "type";
    private static final String INFO = "info";
    private static final String HEALTH = "health";
    private static final String REQUIRED_SERVICES = "mandatory.health.dependencies";
    private static final String OPTIONAL_SERVICES = "optional.health.dependencies";
    private static final String ROUTE = "route";
    private static final String MESSAGE = "message";
    private static final String STATUS = "status";
    private static final String ORIGIN = "origin";
    private static final String NAME = "name";
    private static final String STATUS_CODE = "status_code";
    private static final String REQUIRED = "required";
    private static final String UPSTREAM = "upstream";
    private static final String NOT_FOUND = "not found";
    private static final String PLEASE_CHECK = "Please check - ";
    private final List<String> requiredServices;
    private final List<String> optionalServices;

    public HealthService() {
        AppConfigReader reader = AppConfigReader.getInstance();
        requiredServices = Utility.getInstance().split(reader.getProperty(REQUIRED_SERVICES, ""), ", ");
        if (requiredServices.isEmpty()) {
            log.info("Mandatory service dependencies - {}", requiredServices);
        }
        optionalServices = Utility.getInstance().split(reader.getProperty(OPTIONAL_SERVICES, ""), ", ");
        if (!optionalServices.isEmpty()) {
            log.info("Optional services dependencies - {}", optionalServices);
        }
    }

    @Override
    public Object handleEvent(Map<String, String> headers, Object body, int instance) {
        Platform platform = Platform.getInstance();
        boolean up = true;
        Map<String, Object> result = new HashMap<>();
        /*
         * Checking dependencies
         */
        List<Map<String, Object>> upstream = new ArrayList<>();
        checkServices(upstream, optionalServices, false);
        if (!checkServices(upstream, requiredServices, true)) {
            up = false;
        }
        // checkServices will update the "upstream" service list
        result.put(UPSTREAM, upstream);
        if (upstream.isEmpty()) {
            result.put(MESSAGE, "Did you forget to define "+REQUIRED_SERVICES+" or "+OPTIONAL_SERVICES);
        }
        result.put(STATUS, up? "UP" : "DOWN");
        result.put(ORIGIN, platform.getOrigin());
        result.put(NAME, platform.getName());
        return result;
    }

    @SuppressWarnings("unchecked")
    private boolean checkServices(List<Map<String, Object>> upstream, List<String> healthServices, boolean required) {
        PostOffice po = PostOffice.getInstance();
        boolean up = true;
        for (String route: healthServices) {
            Map<String, Object> m = new HashMap<>();
            m.put(ROUTE, route);
            m.put(REQUIRED, required);
            upstream.add(m);
            try {
                String key = INFO+"/"+route;
                if (!cache.exists(key)) {
                    EventEnvelope infoRes = po.request(route, 3000, new Kv(TYPE, INFO));
                    if (infoRes.getBody() instanceof Map) {
                        cache.put(key, infoRes.getBody());
                    }
                }
                Object info = cache.get(key);
                if (info instanceof Map) {
                    Map<String, Object> map = (Map<String, Object>) info;
                    for (String k : map.keySet()) {
                        m.put(k, map.get(k));
                    }
                }
                EventEnvelope res = po.request(route, 10000, new Kv(TYPE, HEALTH));
                if (res.getBody() instanceof String) {
                    m.put(STATUS_CODE, res.getStatus());
                    m.put(MESSAGE, res.getBody());
                    if (res.getStatus() != 200) {
                        up = false;
                    }
                }
            } catch (IOException e) {
                up = false;
                if (e.getMessage().contains(NOT_FOUND)) {
                    m.put(STATUS_CODE, 404);
                    m.put(MESSAGE, PLEASE_CHECK+e.getMessage());
                } else {
                    m.put(STATUS_CODE, 500);
                    m.put(MESSAGE, e.getMessage());
                }
            } catch (TimeoutException e) {
                up = false;
                m.put(STATUS_CODE, 408);
                m.put(MESSAGE, e.getMessage());
            } catch (AppException e) {
                up = false;
                m.put(STATUS_CODE, e.getStatus());
                m.put(MESSAGE, e.getMessage());
            }
        }
        return up;
    }

}

File: system/platform-core/src/main/kotlin/org/platformlambda/core/actuator/HealthService.kt
/*

    Copyright 2018-2024 Accenture Technology

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.

 */
package org.platformlambda.core.actuator

import org.platformlambda.core.exception.AppException
import org.platformlambda.core.models.EventEnvelope
import org.platformlambda.core.models.KotlinLambdaFunction
import org.platformlambda.core.models.Kv
import org.platformlambda.core.system.FastRPC
import org.platformlambda.core.system.Platform
import org.platformlambda.core.system.EventEmitter
import org.platformlambda.core.util.AppConfigReader
import org.platformlambda.core.util.SimpleCache
import org.platformlambda.core.util.Utility
import org.slf4j.LoggerFactory
import java.io.IOException

class HealthService : KotlinLambdaFunction<EventEnvelope, Any> {
    private val requiredServices: List<String>
    private val optionalServices: List<String>

    init {
        val reader = AppConfigReader.getInstance()
        requiredServices = Utility.getInstance().split(reader.getProperty(REQUIRED_SERVICES, ""), ", ")
        if (requiredServices.isNotEmpty()) {
            log.info("Mandatory service dependencies - {}", requiredServices)
        }
        optionalServices = Utility.getInstance().split(reader.getProperty(OPTIONAL_SERVICES, ""), ", ")
        if (optionalServices.isNotEmpty()) {
            log.info("Optional services dependencies - {}", optionalServices)
        }
    }

    override suspend fun handleEvent(headers: Map<String, String>, input: EventEnvelope, instance: Int): Any {
        val fastRPC = FastRPC(headers)
        val platform = Platform.getInstance()
        val po = EventEmitter.getInstance();
        var up = true
        val result: MutableMap<String, Any> = HashMap()
        /*
         * Checking dependencies
         */
        val upstream: MutableList<Map<String, Any?>> = ArrayList()
        checkServices(fastRPC, upstream, optionalServices, false)
        if (!checkServices(fastRPC, upstream, requiredServices, true)) {
            up = false
        }
        // save the current health status
        po.send(EventEmitter.ACTUATOR_SERVICES, up, Kv(TYPE, HEALTH_STATUS))
        // checkServices will update the "upstream" service list
        result[UPSTREAM] = upstream
        if (upstream.isEmpty()) {
            result[MESSAGE] = "Did you forget to define $REQUIRED_SERVICES or $OPTIONAL_SERVICES"
        }
        result[STATUS] = if (up) "UP" else "DOWN"
        result[ORIGIN] = platform.origin
        result[NAME] = platform.name
        val response = EventEnvelope()
                .setHeader(CONTENT_TYPE, APPLICATION_JSON).setBody(result).setStatus(if (up) 200 else 400)
        val accept = headers[ACCEPT] ?: "?"
        if (accept.startsWith(APPLICATION_XML)) {
            response.setHeader(CONTENT_TYPE, APPLICATION_XML)
        } else {
            response.setHeader(CONTENT_TYPE, APPLICATION_JSON)
        }
        return response
    }

    private suspend fun checkServices(fastRPC: FastRPC, upstream: MutableList<Map<String, Any?>>,
                                      services: List<String>, required: Boolean): Boolean {
        var up = true
        for (route in services) {
            val m: MutableMap<String, Any?> = HashMap()
            m[ROUTE] = route
            m[REQUIRED] = required
            upstream.add(m)
            try {
                val key = "$INFO/$route"
                if (!cache.exists(key)) {
                    val req = EventEnvelope().setTo(route).setHeader(TYPE, INFO)
                    val res = fastRPC.awaitRequest(req, 3000)
                    if (res.body is Map<*, *>) {
                        cache.put(key, res.body)
                    }
                }
                val info = cache[key]
                if (info is Map<*, *>) {
                    for (kv in info) {
                        if (kv.key != null) {
                            m[kv.key.toString()] = kv.value
                        }
                    }
                }
                val req = EventEnvelope().setTo(route).setHeader(TYPE, HEALTH)
                val res = fastRPC.awaitRequest(req, 10000)
                if (res.body is String) {
                    m[STATUS_CODE] = res.status
                    m[MESSAGE] = res.body
                    if (res.status != 200) {
                        up = false
                    }
                }
            } catch (e: IOException) {
                up = false
                if (e.message!!.contains(NOT_FOUND)) {
                    m[STATUS_CODE] = 404
                    m[MESSAGE] = PLEASE_CHECK + e.message
                } else {
                    m[STATUS_CODE] = 500
                    m[MESSAGE] = e.message
                }
            } catch (e: AppException) {
                up = false
                m[STATUS_CODE] = e.status
                m[MESSAGE] = e.message
            }
        }
        return up
    }

    companion object {
        private val log = LoggerFactory.getLogger(HealthService::class.java)
        private val cache = SimpleCache.createCache("health.info", 5000)
        private const val TYPE = "type"
        private const val ACCEPT = "accept"
        private const val INFO = "info"
        private const val HEALTH = "health"
        private const val HEALTH_STATUS = "health_status"
        private const val REQUIRED_SERVICES = "mandatory.health.dependencies"
        private const val OPTIONAL_SERVICES = "optional.health.dependencies"
        private const val ROUTE = "route"
        private const val MESSAGE = "message"
        private const val STATUS = "status"
        private const val ORIGIN = "origin"
        private const val NAME = "name"
        private const val STATUS_CODE = "status_code"
        private const val REQUIRED = "required"
        private const val UPSTREAM = "upstream"
        private const val NOT_FOUND = "not found"
        private const val PLEASE_CHECK = "Please check - "
        private const val CONTENT_TYPE = "content-type"
        private const val APPLICATION_JSON = "application/json"
        private const val APPLICATION_XML = "application/xml"
    }
}